File: src/main/java/org/snake/game/OptionsFrame.java
package org.snake.game;


import java.awt.BorderLayout;
import java.awt.FlowLayout;
import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;

import javax.swing.ButtonGroup;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JRadioButton;
import javax.swing.border.EtchedBorder;

import org.snake.settings.FileManager;
import org.snake.settings.GameConfiguration;

public class OptionsFrame extends JFrame implements ActionListener{
	public static final long serialVersionUID = 1L;
	
	private GameMenu caller;	
	private LabyrinthPreviewer lp;
	
	private JRadioButton snakeSpeedEasy;
	private JRadioButton snakeSpeedMedium;
	private JRadioButton snakeSpeedHard;

	public OptionsFrame(GameMenu caller){
		super("Options");
		this.setResizable(false);		
		
		this.caller = caller;
		this.caller.setVisible(false);
		
		setLayout(new BorderLayout());
		
		//******Choosing Buttons***********
		JButton ok = new JButton("OK");
		JButton cancel = new JButton("Cancel");		
		
		JPanel buttons = new JPanel();				
		buttons.add(ok);
		buttons.add(cancel);		
								
		add(buttons, BorderLayout.SOUTH);				
		
		ok.addActionListener(this);
		cancel.addActionListener(this);			
		
		//*******RadioButtons*****
		snakeSpeedEasy = new JRadioButton("Easy");
		snakeSpeedMedium = new JRadioButton("Medium", true);
		snakeSpeedHard = new JRadioButton("Hard");		
		
		ButtonGroup group = new ButtonGroup();				
		group.add(snakeSpeedEasy);
		group.add(snakeSpeedMedium);
		group.add(snakeSpeedHard);
		
		JPanel radio = new JPanel();
		radio.setLayout(new FlowLayout());
		radio.add(snakeSpeedEasy);
		radio.add(snakeSpeedMedium);
		radio.add(snakeSpeedHard);
		
		JPanel j = new JPanel();
		j.setLayout(new BorderLayout());
		j.setBorder(new EtchedBorder());
		j.add(new JLabel("Snake Speed", JLabel.CENTER), BorderLayout.NORTH);
		j.add(radio, BorderLayout.CENTER);
				
		add(j, BorderLayout.NORTH);
		
		//**LabyrintPreviewer**
		lp = new LabyrinthPreviewer();
		add(lp, BorderLayout.CENTER);		
		
		pack();
		setLocationRelativeTo(null);
		setResizable(false);
		setVisible(true);		
		setDefaultCloseOperation(EXIT_ON_CLOSE);
	}
	
	private void creatGameConfiguration(){
		GameConfiguration g = caller.config;
		
		if(snakeSpeedEasy.isSelected()){
			g.setSnakeSpeedAtEasy();			
		}else{
		if(snakeSpeedMedium.isSelected()){
			g.setSnakeSpeedAtMedium();			
		}else{
		if(snakeSpeedHard.isSelected()){
			g.setSnakeSpeedAtHard();			
		}
		}
		}
		
		g.setLabyrinth(lp.getLabyrinthName());
		
		caller.config = g;					
	}
	
	public void actionPerformed(ActionEvent a){
		String s = a.getActionCommand();			

		if(s.equals("OK")){
			creatGameConfiguration();
			this.dispose();
			caller.setVisible(true);					
		}else{
			if(s.equals("Cancel")){
				this.dispose();
				caller.setVisible(true);
		}
		}
	}

	private class LabyrinthPreviewer extends JPanel implements ActionListener{
		public static final long serialVersionUID = 1L;
			Board b;
			JLabel labyrinthName;
			String[] labyrinths;
			int idx;
		
		public LabyrinthPreviewer(){
			super();
			setLayout(new BorderLayout());
			setBorder(new EtchedBorder());
			
			JButton next = new JButton("Next");			
			JButton previous = new JButton("Previous");
			
			next.addActionListener(this);			
			previous.addActionListener(this);
			
			labyrinthName = new JLabel("NoLabyrinth", JLabel.CENTER);
			labyrinthName.setFont(new Font("Comic Sans MS", Font.BOLD, 14));
			
			add(next, BorderLayout.EAST);
			add(labyrinthName, BorderLayout.CENTER);
			add(previous, BorderLayout.WEST);
						
			b = new Board();
			add(b , BorderLayout.NORTH);
			
			File f = new File("src/main/java/org/snake/files/labyrinths/");
			labyrinths = f.list();
			
			for(int i = 0; i < labyrinths.length; i++){				
				labyrinths[i] = labyrinths[i].substring(0, labyrinths[i].lastIndexOf("."));				
			}
			
			idx = -1;
		}
		
		public String getLabyrinthName(){
			if( idx < 0 || idx >= labyrinths.length ){return null;}
			
			return labyrinths[idx];			
		}
		
		public void actionPerformed(ActionEvent a){
			String action = a.getActionCommand();
			b.clearLabyrinth();
			
			if(action.equals("Next")){
				++idx;
				if(idx >= labyrinths.length){
					idx = -1;
					labyrinthName.setText("No Labyrinth");
				}else{
					String LabyrinthPath = FileManager.LABYRINTH_PATH + labyrinths[idx] + FileManager.LABYRINTH_EXT;
					FileManager.loadLabyrinth(b.getBoard(), LabyrinthPath);
					labyrinthName.setText(labyrinths[idx]);
				}
			}else{				
				if(action.equals("Previous")){
					--idx;
					if(idx < 0){
						idx = labyrinths.length;
						labyrinthName.setText("No Labyrinth");
					}
					else{
						String LabyrinthPath = FileManager.LABYRINTH_PATH + labyrinths[idx] + FileManager.LABYRINTH_EXT;
						FileManager.loadLabyrinth(b.getBoard(), LabyrinthPath);
						labyrinthName.setText(labyrinths[idx]);
					}					
				}
			}
		}
	}
}

File: src/main/kotlin/org/snake/game/OptionsFrame.kt
package org.snake.game

import org.snake.models.GameMatrix
import org.snake.settings.FileManager
import org.snake.settings.SnakeSpeed
import java.awt.BorderLayout
import java.awt.FlowLayout
import java.awt.Font
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.io.File
import javax.swing.*
import javax.swing.border.EtchedBorder

class OptionsFrame(caller: GameMenu) : JFrame("Options"), ActionListener {
    private val caller: GameMenu
    private val lp: LabyrinthPreviewer
    private val snakeSpeedEasy: JRadioButton
    private val snakeSpeedMedium: JRadioButton
    private val snakeSpeedHard: JRadioButton
    private fun creatGameConfiguration() {
        val gameConfiguration = caller.config
        if (snakeSpeedEasy.isSelected) {
            gameConfiguration.snakeSpeed = SnakeSpeed.EASY
        } else if (snakeSpeedMedium.isSelected) {
            gameConfiguration.snakeSpeed = SnakeSpeed.MEDIUM
        } else if (snakeSpeedHard.isSelected) {
            gameConfiguration.snakeSpeed = SnakeSpeed.HARD
        } else {
            gameConfiguration.snakeSpeed = SnakeSpeed.MEDIUM
        }
        gameConfiguration.labyrinthPath = lp.getLabyrinthName()
        caller.config = gameConfiguration!!
    }

    override fun actionPerformed(a: ActionEvent) {
        val s = a.actionCommand
        if (s == "OK") {
            creatGameConfiguration()
            dispose()
            caller.isVisible = true
        } else {
            if (s == "Cancel") {
                dispose()
                caller.isVisible = true
            }
        }
    }

    private inner class LabyrinthPreviewer : JPanel(), ActionListener {
        val cells = GameMatrix(30, 35)
        val board = Board(cells)
        var labyrinthName: JLabel
        var labyrinths: Array<String>
        var idx: Int
        fun getLabyrinthName(): String? {
            return if (idx < 0 || idx >= labyrinths.size) {
                null
            } else labyrinths[idx]
        }

        override fun actionPerformed(a: ActionEvent) {
            val action = a.actionCommand
            board.clearLabyrinth()
            if (action == "Next") {
                ++idx
                if (idx >= labyrinths.size) {
                    idx = -1
                    labyrinthName.text = "No Labyrinth"
                } else {
                    val LabyrinthPath: String = FileManager.Companion.LABYRINTH_PATH + labyrinths[idx] + FileManager.Companion.LABYRINTH_EXT
                    FileManager.Companion.loadLabyrinth(cells, LabyrinthPath)
                    labyrinthName.text = labyrinths[idx]
                }
            } else {
                if (action == "Previous") {
                    --idx
                    if (idx < 0) {
                        idx = labyrinths.size
                        labyrinthName.text = "No Labyrinth"
                    } else {
                        val LabyrinthPath: String = FileManager.Companion.LABYRINTH_PATH + labyrinths[idx] + FileManager.Companion.LABYRINTH_EXT
                        FileManager.Companion.loadLabyrinth(cells, LabyrinthPath)
                        labyrinthName.text = labyrinths[idx]
                    }
                }
            }
            board.repaint()
        }

        init {
            layout = BorderLayout()
            border = EtchedBorder()
            val next = JButton("Next")
            val previous = JButton("Previous")
            next.addActionListener(this)
            previous.addActionListener(this)
            labyrinthName = JLabel("NoLabyrinth", JLabel.CENTER)
            labyrinthName.font = Font("Comic Sans MS", Font.BOLD, 14)
            add(next, BorderLayout.EAST)
            add(labyrinthName, BorderLayout.CENTER)
            add(previous, BorderLayout.WEST)
            add(board, BorderLayout.NORTH)
            val f = File("labyrinths/")
            labyrinths = f.list()
            for (i in labyrinths.indices) {
                labyrinths[i] = labyrinths[i].substring(0, labyrinths[i].lastIndexOf("."))
            }
            idx = -1
        }
    }

    companion object {
        const val serialVersionUID = 1L
    }

    init {
        this.isResizable = false
        this.caller = caller
        this.caller.isVisible = false
        layout = BorderLayout()

        //******Choosing Buttons***********
        val ok = JButton("OK")
        val cancel = JButton("Cancel")
        val buttons = JPanel()
        buttons.add(ok)
        buttons.add(cancel)
        add(buttons, BorderLayout.SOUTH)
        ok.addActionListener(this)
        cancel.addActionListener(this)

        //*******RadioButtons*****
        snakeSpeedEasy = JRadioButton("Easy")
        snakeSpeedMedium = JRadioButton("Medium", true)
        snakeSpeedHard = JRadioButton("Hard")
        val group = ButtonGroup()
        group.add(snakeSpeedEasy)
        group.add(snakeSpeedMedium)
        group.add(snakeSpeedHard)
        val radio = JPanel()
        radio.layout = FlowLayout()
        radio.add(snakeSpeedEasy)
        radio.add(snakeSpeedMedium)
        radio.add(snakeSpeedHard)
        val j = JPanel()
        j.layout = BorderLayout()
        j.border = EtchedBorder()
        j.add(JLabel("Snake Speed", JLabel.CENTER), BorderLayout.NORTH)
        j.add(radio, BorderLayout.CENTER)
        add(j, BorderLayout.NORTH)

        //**LabyrintPreviewer**
        lp = LabyrinthPreviewer()
        add(lp, BorderLayout.CENTER)
        pack()
        setLocationRelativeTo(null)
        isResizable = false
        isVisible = true
        defaultCloseOperation = EXIT_ON_CLOSE
    }
}